pkg/reconciler/pipelinerun/affinity_assistant.go

```diff
@@ -0,0 +1,174 @@
+
+	featureFlagDisableAffinityAssistantKey = "disable-affinity-assistant"
+	affinityAssistantStatefulSetNamePrefix = "affinity-assistant-"
+)
+
+// createAffinityAssistants creates an Affinity Assistant StatefulSet for every workspace in the PipelineRun that
+// use a PersistentVolumeClaim volume. This is done to achieve Node Affinity for all TaskRuns that
+// share the workspace volume and make it possible for the tasks to execute parallel while sharing volume.
+func (c *Reconciler) createAffinityAssistants(wb []v1alpha1.WorkspaceBinding, ownerReference metav1.OwnerReference, namespace string) error {
+	var errs []error
+	for _, w := range wb {
+		if w.PersistentVolumeClaim != nil || w.VolumeClaimTemplate != nil {
+			affinityAssistantName := affinityAssistantName(w.Name, ownerReference)
+			_, err := c.KubeClientSet.AppsV1().StatefulSets(namespace).Get(affinityAssistantName, metav1.GetOptions{})
+			switch {
+			case apierrors.IsNotFound(err):
+				_, err := c.KubeClientSet.AppsV1().StatefulSets(namespace).Create(affinityAssistantStatefulSet(affinityAssistantName, ownerReference))
+				if err != nil {
+					errs = append(errs, fmt.Errorf("failed to create StatefulSet %s: %s", affinityAssistantName, err))
+				}
+				if err == nil {
+					c.Logger.Infof("Created StatefulSet %s in namespace %s", affinityAssistantName, namespace)
+				}
+			case err != nil:
+				errs = append(errs, fmt.Errorf("failed to retrieve StatefulSet %s: %s", affinityAssistantName, err))
+			}
+		}
+	}
+	return errorutils.NewAggregate(errs)
+}
+
+func (c *Reconciler) cleanupAffinityAssistants(pr *v1beta1.PipelineRun) error {
+	var errs []error
+	for _, w := range pr.Spec.Workspaces {
+		if w.PersistentVolumeClaim != nil || w.VolumeClaimTemplate != nil {
+			affinityAssistantStsName := affinityAssistantStatefulSetNamePrefix + affinityAssistantName(w.Name, pr.GetOwnerReference())
+			if err := c.KubeClientSet.AppsV1().StatefulSets(pr.Namespace).Delete(affinityAssistantStsName, &metav1.DeleteOptions{}); err != nil {
+				errs = append(errs, fmt.Errorf("failed to delete StatefulSet %s: %s", affinityAssistantStsName, err))
+			}
+		}
+	}
+	return errorutils.NewAggregate(errs)
+}
+
+func affinityAssistantName(pipelineWorkspaceName string, owner metav1.OwnerReference) string {
+	return fmt.Sprintf("%s-%s", pipelineWorkspaceName, owner.Name)
+}
+
+func affinityAssistantStatefulSet(name string, ownerReference metav1.OwnerReference) *appsv1.StatefulSet {
+	// We want a singleton pod
+	replicas := int32(1)
+
+	// LabelInstance is used to configure PodAffinity for all TaskRuns belonging to this Affinity Assistant
+	// LabelComponent is used to configure PodAntiAffinity to other Affinity Assistants
+	labels := map[string]string{
+		workspace.LabelInstance:  name,
+		workspace.LabelComponent: workspace.ComponentNameAffinityAssistant,
+	}
+
+	containers := []corev1.Container{{
+		Name: "affinity-assistant",
+
+		// We may want to create a custom, minimal binary
```

Comment:
Can you make this a TODO with an issue number? We almost certainly want a smaller purpose-built binary for this, and luckily it won't be that hard to add.